File: LeetCode/LeetCode-CN-CPP/2020六月每日一题/1014. 最佳观光组合.cpp
class Solution {
public:
    int maxScoreSightseeingPair(vector<int>& A) {
        // 将A[i] + A[j] + i - j 分解为：A[j] - j + A[i] + i
        int ans = 0;
        int mx = A[0] + 0; // 记录最大的A[i] + i;
        for(int i = 1; i < A.size(); i++) {
            mx = max(mx, A[i - 1] + i - 1); // 记录最大的A[i] + i
            ans = max(ans, A[i] - i + mx); // 针对当前j来说，计算最大的A[i] + i + A[j] - j
        }
        return ans;
    }
};

File: LeetCode/LeetCode-CN-JAVA/2020六月每日一题/1014. 最佳观光组合.java
class Solution {
    public int maxScoreSightseeingPair(int[] A) {
        int ans = 0;
        int mx = A[A.length - 1] - A.length - 1; // 记录最大的 A[j] - j
        for(int i = A.length - 2; i >= 0; i--) {
            mx = Math.max(mx, A[i + 1] - i - 1); // 计算出最大的A[j] - j
            ans = Math.max(ans, mx + A[i] + i);
        }
        return ans;
    }
}